amazon/ion/ioncmodule.c

```diff
@@ -0,0 +1,1369 @@
+
+    if (PyUnicode_Check(obj)) {
+        if (ion_type == tid_none_INT) {
+            ion_type = tid_STRING_INT;
+        }
+        ION_STRING string_value;
+        ion_string_from_py(obj, &string_value);
+        if (tid_STRING_INT == ion_type) {
+            IONCHECK(ion_writer_write_string(writer, &string_value));
+        }
+        else if (tid_SYMBOL_INT == ion_type) {
+            IONCHECK(ion_writer_write_symbol(writer, &string_value));
+        }
+        else {
+            _FAILWITHMSG(IERR_INVALID_ARG, "Found text; expected STRING or SYMBOL Ion type.");
+        }
+    }
+    else if (PyBool_Check(obj)) { // NOTE: this must precede the INT block because python bools are ints.
+        if (ion_type == tid_none_INT) {
+            ion_type = tid_BOOL_INT;
+        }
+        if (tid_BOOL_INT != ion_type) {
+            _FAILWITHMSG(IERR_INVALID_ARG, "Found bool; expected BOOL Ion type.");
+        }
+        BOOL bool_value;
+        if (obj == Py_True) {
+            bool_value = TRUE;
+        }
+        else {
+            bool_value = FALSE;
+        }
+        IONCHECK(ion_writer_write_bool(writer, bool_value));
+    }
+    else if (PyInt_Check(obj)) {
+        if (ion_type == tid_none_INT) {
+            ion_type = tid_INT_INT;
+        }
+        if (tid_INT_INT == ion_type) {
+            IONCHECK(ionc_write_big_int(writer, obj));
+        }
+        else if (tid_BOOL_INT == ion_type) {
+            IONCHECK(ion_writer_write_bool(writer, PyInt_AsSsize_t(obj)));
+        }
+        else {
+            _FAILWITHMSG(IERR_INVALID_ARG, "Found int; expected INT or BOOL Ion type.");
+        }
+    }
+    else if (PyFloat_Check(obj)) {
+        if (ion_type == tid_none_INT) {
+            ion_type = tid_FLOAT_INT;
+        }
+        if (tid_FLOAT_INT != ion_type) {
+            _FAILWITHMSG(IERR_INVALID_ARG, "Found float; expected FLOAT Ion type.");
+        }
+        IONCHECK(ion_writer_write_double(writer, PyFloat_AsDouble(obj)));
+    }
+    else if (PyObject_TypeCheck(obj, (PyTypeObject*)_ionpynull_cls)) {
+        if (ion_type == tid_none_INT) {
+            ion_type = tid_NULL_INT;
+        }
+        IONCHECK(ion_writer_write_typed_null(writer, (ION_TYPE)ion_type));
+    }
+    else if (PyObject_TypeCheck(obj, (PyTypeObject*)_decimal_constructor)) {
+        if (ion_type == tid_none_INT) {
+            ion_type = tid_DECIMAL_INT;
+        }
+        if (tid_DECIMAL_INT != ion_type) {
+            _FAILWITHMSG(IERR_INVALID_ARG, "Found Decimal; expected DECIMAL Ion type.");
+        }
+
+        PyObject* decimal_str = PyObject_CallMethod(obj, "__str__", NULL); // TODO converting every decimal to string is slow.
+        char* decimal_c_str = NULL;
+        Py_ssize_t decimal_c_str_len;
+        IONCHECK(c_string_from_py(decimal_str, &decimal_c_str, &decimal_c_str_len));
+
+        ION_DECIMAL decimal_value;
+        IONCHECK(ion_decimal_from_string(&decimal_value, decimal_c_str, &dec_context));
```

Comment:
Replaced previous `decQuadFromString` to `ion_decimal_from_string`.

Refer to [here](https://github.com/cheqianh/ion-python/blob/43bf9243c9b145f8e2baca32624f693bfb3d87ab/amazon/ion/ioncmodule.c#L371) for previous code.

Reply:
As discussed offline, we only support 34 decimal digits. There is a test with more than 34 digits throws NUMBER_FLOW exception. 

```
_IT.DECIMAL: (
(_D('1.1999999999999999555910790149937383830547332763671875e0'),   # <- value
 b'1.1999999999999999555910790149937383830547332763671875'))           # <- expected value
```

dump/write decimal function [here](https://github.com/amzn/ion-python/pull/149/commits/a176f6f483191aa6c9907dc4e66af22bbd912af9#diff-535412a8258ff8b32325ed24da1018a8b331ec818afbda63f3d585d631c5847dR633).
load/read decimal function [here](https://github.com/amzn/ion-python/pull/149/commits/a176f6f483191aa6c9907dc4e66af22bbd912af9#diff-535412a8258ff8b32325ed24da1018a8b331ec818afbda63f3d585d631c5847dR1181).